lightning/src/ln/peers/handshake/mod.rs

```diff
@@ -392,3 +392,374 @@ impl PeerHandshake {
+		assert_eq!(test_ctx.outbound_handshake.get_remote_pubkey(), Some(test_ctx.inbound_public_key));
+	}
+
+
+	// Outbound::AwaitingActTwo -> Complete (with extra data)
+	// Ensures that any remaining data in the read buffer is transferred to the conduit once
+	// the handshake is complete
+	// TODO: Is this valid? Don't we expect peers to need ActThree before sending additional data?
+	#[test]
+	fn peer_handshake_new_outbound_excess_bytes_after_complete_are_in_conduit() {
+		let mut test_ctx = TestCtx::new();
+		let act1 = do_process_act_or_panic!(test_ctx.outbound_handshake, &[]);
+		let mut act2 = do_process_act_or_panic!(test_ctx.inbound_handshake, &act1);
+		act2.extend_from_slice(&[1; 100]);
+
+		let conduit = if let (_, Some(conduit)) = test_ctx.outbound_handshake.process_act(&act2).unwrap() {
+			conduit
+		} else {
+			panic!();
+		};
+
+		assert_eq!(100, conduit.decryptor.read_buffer_length());
+		assert_eq!(test_ctx.outbound_handshake.get_remote_pubkey(), Some(test_ctx.inbound_public_key));
+	}
+
+	// Outbound::AwaitingActTwo -> Error (input too small)
+	#[test]
+	fn peer_handshake_outbound_awaiting_act_two_input_too_small() {
+		let mut test_ctx = TestCtx::new();
+		let _act1 = do_process_act_or_panic!(test_ctx.outbound_handshake, &[]);
+
+		assert_eq!(test_ctx.outbound_handshake.process_act(&[1]).err(), Some(String::from("need at least 50 bytes")));
+	}
+
+	// Outbound::AwaitingActTwo -> Error (bad version byte)
+	#[test]
+	fn peer_handshake_outbound_awaiting_act_two_bad_version() {
+		let mut test_ctx = TestCtx::new();
+		let act1 = do_process_act_or_panic!(test_ctx.outbound_handshake, &[]);
+		let mut act2 = do_process_act_or_panic!(test_ctx.inbound_handshake, &act1);
+		// Set version byte to 1
+		act2[0] = 1;
+
+		assert_eq!(test_ctx.outbound_handshake.process_act(&act2).err(), Some(String::from("unexpected version")));
+	}
+
+	// Outbound::AwaitingActTwo -> Error (invalid hmac)
+	#[test]
+	fn peer_handshake_outbound_awaiting_act_two_invalid_hmac() {
+		let mut test_ctx = TestCtx::new();
+		let act1 = do_process_act_or_panic!(test_ctx.outbound_handshake, &[]);
+		let mut act2 = do_process_act_or_panic!(test_ctx.inbound_handshake, &act1);
+		// corrupt the ciphertext
+		act2[34] = 1;
+
+		assert_eq!(test_ctx.outbound_handshake.process_act(&act2).err(), Some(String::from("invalid hmac")));
+	}
+
+	// Outbound::AwaitingActTwo -> Error (invalid remote ephemeral key)
+	#[test]
+	fn peer_handshake_outbound_awaiting_act_two_invalid_remote_ephemeral_key() {
+		let mut test_ctx = TestCtx::new();
+		let act1 = do_process_act_or_panic!(test_ctx.outbound_handshake, &[]);
+		let mut act2 = do_process_act_or_panic!(test_ctx.inbound_handshake, &act1);
+		// corrupt the ephemeral public key
+		act2[1] = 1;
+
+		assert_eq!(test_ctx.outbound_handshake.process_act(&act2).err(), Some(String::from("invalid remote ephemeral public key")));
+	}
+
+	// Inbound::AwaitingActThree -> None
+	// TODO: should this transition to Complete instead of None?
```

Comment:
Likely it should. @arik-so?

Reply:
As discussed in Slack. It is probably best to view this first PR in the "files changed" view instead of commit-by-commit. Otherwise, you will see all of the bugs from 494 before they were fixed in subsequent commits.